### PR TITLE
Fix BSONWalkAll unexpected REMOVE_FIELD behavior

### PR DESCRIPTION
**Pull request by NameThatFollows**
Running `BSONWalkAll` with a visitor that returns a `REMOVE_FIELD` resulted in some unexpected behavior:

- The last element in the resulting BSON would be duplicated (new slice was not assigned to the element, so existing slice would still show the same number of elements)
- Removing a field would cause the walk to short-circuit, preventing candidates in further iterations from being seen. 

This also adds a test to cover these two cases.

### CODE REVIEW

**Comment by TomerYakir** [comment]
LGTM - adding @d-kupiec for another pair of eyes

**Review by TomerYakir — APPROVED** [review]

File: bsonutil.go [hunk 1]
```diff
@@ -209,10 +210,13 @@ func BSONWalkAll(doc bson.D, fieldName string, visitor BSONWalkVisitor) (bson.D,
 				if err == DELETE_ME {
 					return nil, nil
```
Comment (d-kupiec) [inline bsonutil.go:211]: By similar logic, does DELETE_ME delete the entire given bson or just the element in the current doc that generated DELETE_ME on Visit()?

File: bsonutil.go [hunk 2]
```diff
@@ -209,10 +210,13 @@ func BSONWalkAll(doc bson.D, fieldName string, visitor BSONWalkVisitor) (bson.D,
...
 				}
+				// if one field needs to be removed, store the index, and continue processing remaining
+				// fields/documents
 				if err == REMOVE_FIELD {
-					return append(current[:i], current[i+1:]...), nil
+					docRemovedIndex = i
```
Comment (d-kupiec) [inline bsonutil.go:216]: If the field is marked for removal, perhaps we should still short circuit the processing below and continue to the next element

File: bsonutil.go [hunk 3]
```diff
@@ -254,6 +258,9 @@ func BSONWalkAll(doc bson.D, fieldName string, visitor BSONWalkVisitor) (bson.D,
 		}
 
 	}
+	if docRemovedIndex > -1 {
```
Comment (d-kupiec) [inline bsonutil.go:261]: Do we expect to only encounter one element resulting in REMOVE_FIELD per call, or might we see more than one now that we arent preemptively returning on the first one?

**Review by d-kupiec — COMMENTED** [review]
a few questions as I learn more about this particular area of code and the expectations for it

File: bsonutil.go (hunk 1 above)
Comment (NameThatFollows) [reply bsonutil.go:211]: I believe the former, DELETE_ME deletes the entire bson that contains the field.

Comment (d-kupiec) [reply bsonutil.go:211]: Got it! Would you mind updating the comment above to rephrase from 'that element' to 'the BSON doc' for clarity?

File: bsonutil.go (hunk 3 above)
Comment (NameThatFollows) [reply bsonutil.go:261]: I assumed that a document (excluding sub-documents) will only ever have one instance of a given field/key. Is it a valid use-case to have a bson with multiple fields with the same field name? If so (or just to be safe), I can probably modify this to track multiple removed fields.

File: bsonutil.go (hunk 2 above)
Comment (NameThatFollows) [reply bsonutil.go:216]: Good point! Let me try out that change

File: bsonutil.go (hunk 3 above)
Comment (d-kupiec) [reply bsonutil.go:261]: It looks like for BSON, it is possible a document may have more than one field with the same name (though unlikely), https://www.mongodb.com/docs/manual/core/document/#field-names

Not sure if Tomer has some more insight regarding handling of this scenario, though I imagine for a base level util we probably want to support it if our docs specify it's allowed

File: bsonutil.go [hunk 4]
```diff
@@ -198,29 +198,29 @@ type BSONWalkVisitor interface {
 }
 
 // BSONWalkAll - recursively walks the BSON doc and applies the visitor when encountering the fieldName
-// If delete_me is encountered, it'll return an empty document for that element
+// If delete_me is encountered, it'll return an empty document for that BSON doc
+// If remove_field is encountered, it'll return the BSON doc with the field ommitted
```
Comment (d-kupiec) [inline bsonutil.go:202]: nit: s/ommitted/omitted

File: bsonutil.go [hunk 5]
```diff
@@ -258,10 +258,11 @@ func BSONWalkAll(doc bson.D, fieldName string, visitor BSONWalkVisitor) (bson.D,
 		}
 
 	}
-	if docRemovedIndex > -1 {
-		return append(current[:docRemovedIndex], current[docRemovedIndex+1:]...), nil
+	for i := len(docsToRemove) - 1; i >= 0; i-- {
+		current = append(current[:i], current[i+1:]...)
```
Comment (d-kupiec) [inline bsonutil.go:262]: I think we want docsToRemove[i]

File: bsonutil_test.go [hunk 6]
```diff
@@ -514,6 +526,57 @@ func TestBSONWalkAll7(test *testing.T) {
...
+	}
+	walker := &removeWalker{}
+	doc, err := BSONWalkAll(doc, "a", walker)
+	if err != nil {
+		test.Errorf("why did we get an error %s", err)
+	}
+	if len(walker.seen) != 2 {
+		test.Errorf("wrong # saw %d", len(walker.seen))
+	}
+	if len(doc) != 1 {
```
Comment (d-kupiec) [inline bsonutil_test.go:543]: Perhaps assert the remaining element is b to ensure the right elements are being removed.

File: bsonutil.go (hunk 5 above)
Comment (NameThatFollows) [reply bsonutil.go]: 🤦 thanks for catching that!

**Review by d-kupiec — APPROVED** [review]
lgtm, thanks!

**Pull request merged**